core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -69,36 +70,61 @@ public void start() throws Exception {
       throw new RuntimeException(e);
     }
 
+    startMasters(false);
     while (!Thread.interrupted()) {
-      startMasters(false);
-      LOG.info("Secondary started");
       mLeaderSelector.waitForState(State.PRIMARY);
+      if (gainPrimacy()) {
+        mLeaderSelector.waitForState(State.SECONDARY);
+        losePrimacy();
+      }
+    }
+  }
+
+  /**
+   * @return true if the master successfully upgrades to primary
```

Comment:
I don't fully understand what this method does. Has the `mLeaderSelector` already notified this should be primary now? This method just does the work to transition?

Reply:
Right, the main loop is
```java
while (!Thread.interrupted()) {
      mLeaderSelector.waitForState(State.PRIMARY);
      if (gainPrimacy()) {
        mLeaderSelector.waitForState(State.SECONDARY);
        losePrimacy();
      }
    }
```
so we only call this method after the leader selector decides that it is primary, and this method does the work to transition. I've updated the javadoc and added a comment to give more explanation